In [1]:
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
%matplotlib inline

In [2]:
df = pd.read_csv('UsedVehicles_with_state_names.csv')

In [3]:
df.head(2)

id                                                url  \
0  7184791621  https://duluth.craigslist.org/ctd/d/duluth-200...   
1  7184773187  https://duluth.craigslist.org/cto/d/saginaw-20...   

              region                     region_url  price    year  \
0  duluth / superior  https://duluth.craigslist.org   6995  2000.0   
1  duluth / superior  https://duluth.craigslist.org   8750  2013.0   

  manufacturer            model  condition    cylinders  ... drive  size type  \
0          gmc  new sierra 1500  excellent  8 cylinders  ...   4wd   NaN  NaN   
1      hyundai           sonata  excellent  4 cylinders  ...   fwd   NaN  NaN   

  paint_color                                          image_url  \
0         red  https://images.craigslist.org/00n0n_f06ykBMcdh...   
1        grey  https://images.craigslist.org/00d0d_kgZ6xoeRw2...   

                                         description county      lat     long  \
0  2000 *** GMC New Sierra 1500 Ext Cab 157.5  WB...    NaN  46.8433 -92.2550   
1  For Sale:  2013 Hyundai Sonata GLS - $8750.  O...    NaN  46.9074 -92.4638   

       State  
0  Minnesota  
1  Minnesota  

[2 rows x 25 columns]

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 423857 entries, 0 to 423856
Data columns (total 25 columns):
 #   Column        Non-Null Count   Dtype  
---  ------        --------------   -----  
 0   id            423857 non-null  int64  
 1   url           423857 non-null  object 
 2   region        423857 non-null  object 
 3   region_url    423857 non-null  object 
 4   price         423857 non-null  int64  
 5   year          328743 non-null  float64
 6   manufacturer  313242 non-null  object 
 7   model         325384 non-null  object 
 8   condition     176719 non-null  object 
 9   cylinders     197679 non-null  object 
 10  fuel          327214 non-null  object 
 11  odometer      270585 non-null  float64
 12  title_status  327759 non-null  object 
 13  transmission  328065 non-null  object 
 14  vin           184420 non-null  object 
 15  drive         231119 non-null  object 
 16  size          102627 non-null  object 
 17  type          241157 non-null  object 
 18  pain

In [5]:
df['vin'].value_counts()

1HTMRAAM19J040080    133
1FVDBSDC34HM88268    125
1FVDCXBS48HZ75358    118
77777777777777777     70
2NKHHJ7X2HM136714     60
                    ... 
2G1WT58K789144286      1
3C7WRVKG9JE152412      1
5UXWZ7C37H0V89377      1
1FTEW1EP3GKE26263      1
2A4RR2D12AR357273      1
Name: vin, Length: 103588, dtype: int64

It looks like some vehicles are represented more than once as vehicle identification numbers are unique to each vehicle

In [6]:
df['vin'].value_counts().count()

103588

In [7]:
#check for duplicate entries in the vin column
vinduplicated = df['vin'].duplicated().value_counts()
vinduplicated

True     320268
False    103589
Name: vin, dtype: int64

There are only 103,588 unique values of 'vin'. Some of these values are missing values, let's see how many missing values we have in the vin column

In [8]:
df['vin'].isnull().sum()

239437

There're a total of 423,857  rows in the vehicles dataframe, out of which 56.49% (239,437) of vehicles identification numbers (vin) are not entered. This means we should have (423,857 - 239,437) = 184,420 unique vin's since no two vehicles can have the same vin,  but on the contrary, there are only 103,588 unique vin's, which means 80,832 vin's are duplicated. 


We will keep the 239,437 rows with null 'vin' values in a separate dataframe, drop vin duplicates and then concatenate both dataframes. The resulting dataframe will have (239,437 + 103,588)  = 343,025 rows

In [9]:
null_vin = df[df['vin'].isnull()]
null_vin

id                                                url  \
1       7184773187  https://duluth.craigslist.org/cto/d/saginaw-20...   
16      7184663608  https://duluth.craigslist.org/cto/d/moose-lake...   
17      7184626883  https://duluth.craigslist.org/ctd/d/barnum-199...   
18      7184618888  https://duluth.craigslist.org/cto/d/duluth-toy...   
21      7184513461  https://duluth.craigslist.org/ctd/d/saint-paul...   
...            ...                                                ...   
423848  7190868079  https://terrehaute.craigslist.org/cto/d/rockvi...   
423849  7190804455  https://terrehaute.craigslist.org/ctd/d/kokomo...   
423850  7190719718  https://terrehaute.craigslist.org/cto/d/terre-...   
423851  7190679541  https://terrehaute.craigslist.org/cto/d/charle...   
423852  7190671647  https://terrehaute.craigslist.org/cto/d/rockvi...   

                   region                         region_url  price    year  \
1       duluth / superior      https://duluth.craigslist.org   8750  2013.0   
16      duluth / superior      https://duluth.craigslist.org   1500  1994.0   
17      duluth / superior      https://duluth.craigslist.org   6000  1997.0   
18      duluth / superior      https://duluth.craigslist.org   3400  2007.0   
21      duluth / superior      https://duluth.craigslist.org  13995  2011.0   
...                   ...                                ...    ...     ...   
423848        terre haute  https://terrehaute.craigslist.org   3900  2006.0   
423849        terre haute  https://terrehaute.craigslist.org  19995  2016.0   
423850        terre haute  https://terrehaute.craigslist.org   1400  2002.0   
423851        terre haute  https://terrehaute.craigslist.org   4500  2006.0   
423852        terre haute  https://terrehaute.craigslist.org   4400  1996.0   

       manufacturer             model  condition    cylinders  ... drive  \
1           hyundai            sonata  excellent  4 cylinders  ...   fwd   
16            buick           century       fair  4 cylinders  ...   fwd   
17             ford             f-800       good          NaN  ...   rwd   
18           toyota        corolla le       good  4 cylinders  ...   fwd   
21             jeep    grand cherokee        NaN          NaN  ...   NaN   
...             ...               ...        ...          ...  ...   ...   
423848         ford           f150 xl       good  6 cylinders  ...   rwd   
423849         mini            cooper        NaN          NaN  ...   4wd   
423850      mercury  sable ls premium       fair  6 cylinders  ...   fwd   
423851       toyota         sienna le  excellent  6 cylinders  ...   fwd   
423852         ford           f250 xl       fair  8 cylinders  ...   4wd   

             size      type paint_color  \
1             NaN       NaN        grey   
16       mid-size     sedan       brown   
17      full-size     other       brown   
18            NaN       NaN       black   
21            NaN       NaN         NaN   
...           ...       ...         ...   
423848  full-size    pickup       brown   
423849        NaN       NaN         NaN   
423850        NaN     sedan         NaN   
423851  full-size  mini-van       brown   
423852  full-size    pickup       white   

                                                image_url  \
1       https://images.craigslist.org/00d0d_kgZ6xoeRw2...   
16      https://images.craigslist.org/00w0w_gOB1mPM5GJ...   
17      https://images.craigslist.org/00x0x_lJHXOnjy35...   
18      https://images.craigslist.org/00T0T_hdofjvggWV...   
21      https://images.craigslist.org/00d0d_eljQ7PhV4B...   
...                                                   ...   
423848  https://images.craigslist.org/00J0J_cDy3JSNnQh...   
423849  https://images.craigslist.org/00c0c_5US6NwjpNx...   
423850  https://images.craigslist.org/00707_1IB2ZGpBWf...   
423851  https://images.craigslist.org/00606_isQJg8sRsW...   
423852  https://images.craigslist.org/00707_8KE0n5KqGK...   

                           

In [10]:
#check for duplicate entries in the vin column after dropping null values
vinduplicated = df['vin'].dropna().duplicated(keep = 'first').value_counts()
vinduplicated

False    103588
True      80832
Name: vin, dtype: int64

In [11]:
vin_df = df[df['vin'].notnull()]
vin_df

id                                                url  \
0       7184791621  https://duluth.craigslist.org/ctd/d/duluth-200...   
2       7184712241  https://duluth.craigslist.org/ctd/d/rush-city-...   
3       7184711787  https://duluth.craigslist.org/ctd/d/rush-city-...   
4       7184710887  https://duluth.craigslist.org/ctd/d/rush-city-...   
5       7184710023  https://duluth.craigslist.org/ctd/d/rush-city-...   
...            ...                                                ...   
423844  7191675129  https://terrehaute.craigslist.org/ctd/d/terre-...   
423853  7190647600  https://terrehaute.craigslist.org/ctd/d/terre-...   
423854  7190623809  https://terrehaute.craigslist.org/ctd/d/terre-...   
423855  7190617178  https://terrehaute.craigslist.org/ctd/d/terre-...   
423856  7190607038  https://terrehaute.craigslist.org/ctd/d/terre-...   

                   region                         region_url  price    year  \
0       duluth / superior      https://duluth.craigslist.org   6995  2000.0   
2       duluth / superior      https://duluth.craigslist.org  16995  2007.0   
3       duluth / superior      https://duluth.craigslist.org  13995  2012.0   
4       duluth / superior      https://duluth.craigslist.org   7995  2010.0   
5       duluth / superior      https://duluth.craigslist.org   8995  2011.0   
...                   ...                                ...    ...     ...   
423844        terre haute  https://terrehaute.craigslist.org  19500  2018.0   
423853        terre haute  https://terrehaute.craigslist.org  17500  2015.0   
423854        terre haute  https://terrehaute.craigslist.org  29500  2013.0   
423855        terre haute  https://terrehaute.craigslist.org  38500  2016.0   
423856        terre haute  https://terrehaute.craigslist.org  32000  2019.0   

       manufacturer                  model  condition    cylinders  ... drive  \
0               gmc        new sierra 1500  excellent  8 cylinders  ...   4wd   
2               gmc  sierra classic 2500hd       good  8 cylinders  ...   4wd   
3              ford                  f-150       good  6 cylinders  ...   4wd   
4         chevrolet                equinox       good  4 cylinders  ...   4wd   
5         chevrolet               traverse       good  6 cylinders  ...   4wd   
...             ...                    ...        ...          ...  ...   ...   
423844       toyota                  camry       good  4 cylinders  ...   NaN   
423853         ford                 escape       good  4 cylinders  ...   NaN   
423854    chevrolet              avalanche       good  8 cylinders  ...   NaN   
423855          gmc          yukon xl 1500       good  8 cylinders  ...   NaN   
423856     chrysler               pacifica       good  6 cylinders  ...   NaN   

             size       type paint_color  \
0             NaN        NaN         red   
2       full-size      truck       white   
3       full-size      truck        grey   
4       full-size        SUV        grey   
5       full-size        SUV       white   
...           ...        ...         ...   
423844        NaN      sedan      silver   
423853        NaN  hatchback        grey   
423854        NaN      other       white   
423855        NaN  hatchback       white   
423856        NaN        van      silver   

                                                image_url  \
0       https://images.craigslist.org/00n0n_f06ykBMcdh...   
2       https://images.craigslist.org/01414_g093aPtSMW...   
3       https://images.craigslist.org/00W0W_io2cwEVtyh...   
4       https://images.craigslist.org/00404_j5GItixnjX...   
5       https://images.craigslist.org/00T0T_bwYCuKicjX...   
...                                                   ...   
423844  https://images.craigslist.org/01616_idlKjmGFKF...   
423853  https://images.craigslist.org/00L0L_gUkOAFGCVr...   
423854  https://images.craigslist.org/00a0a_aSnYnC5gJS...   
423855  https://images.craigslist.org/00808_1zOiRRUAl4...   
423856  https://i

In [12]:
#Drop duplicated columns from vin_df
vins = vin_df['vin']
vin_unique = vin_df.drop_duplicates(subset = ['vin'])
vin_unique

id                                                url  \
0       7184791621  https://duluth.craigslist.org/ctd/d/duluth-200...   
2       7184712241  https://duluth.craigslist.org/ctd/d/rush-city-...   
3       7184711787  https://duluth.craigslist.org/ctd/d/rush-city-...   
4       7184710887  https://duluth.craigslist.org/ctd/d/rush-city-...   
5       7184710023  https://duluth.craigslist.org/ctd/d/rush-city-...   
...            ...                                                ...   
423821  7191703033  https://southbend.craigslist.org/ctd/d/otsego-...   
423837  7192249724  https://terrehaute.craigslist.org/ctd/d/terre-...   
423854  7190623809  https://terrehaute.craigslist.org/ctd/d/terre-...   
423855  7190617178  https://terrehaute.craigslist.org/ctd/d/terre-...   
423856  7190607038  https://terrehaute.craigslist.org/ctd/d/terre-...   

                       region                         region_url  price  \
0           duluth / superior      https://duluth.craigslist.org   6995   
2           duluth / superior      https://duluth.craigslist.org  16995   
3           duluth / superior      https://duluth.craigslist.org  13995   
4           duluth / superior      https://duluth.craigslist.org   7995   
5           duluth / superior      https://duluth.craigslist.org   8995   
...                       ...                                ...    ...   
423821  south bend / michiana   https://southbend.craigslist.org  18990   
423837            terre haute  https://terrehaute.craigslist.org  19000   
423854            terre haute  https://terrehaute.craigslist.org  29500   
423855            terre haute  https://terrehaute.craigslist.org  38500   
423856            terre haute  https://terrehaute.craigslist.org  32000   

          year manufacturer                  model  condition    cylinders  \
0       2000.0          gmc        new sierra 1500  excellent  8 cylinders   
2       2007.0          gmc  sierra classic 2500hd       good  8 cylinders   
3       2012.0         ford                  f-150       good  6 cylinders   
4       2010.0    chevrolet                equinox       good  4 cylinders   
5       2011.0    chevrolet               traverse       good  6 cylinders   
...        ...          ...                    ...        ...          ...   
423821  2017.0         ford                 fusion        NaN  4 cylinders   
423837  2019.0         ford               ecosport       good  4 cylinders   
423854  2013.0    chevrolet              avalanche       good  8 cylinders   
423855  2016.0          gmc          yukon xl 1500       good  8 cylinders   
423856  2019.0     chrysler               pacifica       good  6 cylinders   

        ... drive       size       type paint_color  \
0       ...   4wd        NaN        NaN         red   
2       ...   4wd  full-size      truck       white   
3       ...   4wd  full-size      truck        grey   
4       ...   4wd  full-size        SUV        grey   
5       ...   4wd  full-size        SUV       white   
...     ...   ...        ...        ...         ...   
423821  ...   fwd        NaN      sedan       white   
423837  ...   NaN        NaN  hatchback        blue   
423854  ...   NaN        NaN      other       white   
423855  ...   NaN        NaN  hatchback       white   
423856  ...   NaN        NaN        van      silver   

                                                image_url  \
0       https://images.craigslist.org/00n0n_f06ykBMcdh...   
2       https://images.craigslist.org/01414_g093aPtSMW...   
3       https://images.craigslist.org/00W0W_io2cwEVtyh...   
4       https://images.craigslist.org/00404_j5GItixnjX...   
5       https://images.craigslist.org/00T0T_bwYCuKicjX...   
...                                                   ...   
423821  https://images.craigslist.org/00R0R_5p4FAvPAyc...   
423837  https://images.craigslist.org/00808_aI3CWupjCT...   
423854  https://images.craigslist.org/00a0a_aSnYnC5gJS...   
423855  https://images.craigsl

In [13]:
#Concat vin_unique with null_vin
vehicles_df = pd.concat([vin_unique, null_vin])
vehicles_df

id                                                url  \
0       7184791621  https://duluth.craigslist.org/ctd/d/duluth-200...   
2       7184712241  https://duluth.craigslist.org/ctd/d/rush-city-...   
3       7184711787  https://duluth.craigslist.org/ctd/d/rush-city-...   
4       7184710887  https://duluth.craigslist.org/ctd/d/rush-city-...   
5       7184710023  https://duluth.craigslist.org/ctd/d/rush-city-...   
...            ...                                                ...   
423848  7190868079  https://terrehaute.craigslist.org/cto/d/rockvi...   
423849  7190804455  https://terrehaute.craigslist.org/ctd/d/kokomo...   
423850  7190719718  https://terrehaute.craigslist.org/cto/d/terre-...   
423851  7190679541  https://terrehaute.craigslist.org/cto/d/charle...   
423852  7190671647  https://terrehaute.craigslist.org/cto/d/rockvi...   

                   region                         region_url  price    year  \
0       duluth / superior      https://duluth.craigslist.org   6995  2000.0   
2       duluth / superior      https://duluth.craigslist.org  16995  2007.0   
3       duluth / superior      https://duluth.craigslist.org  13995  2012.0   
4       duluth / superior      https://duluth.craigslist.org   7995  2010.0   
5       duluth / superior      https://duluth.craigslist.org   8995  2011.0   
...                   ...                                ...    ...     ...   
423848        terre haute  https://terrehaute.craigslist.org   3900  2006.0   
423849        terre haute  https://terrehaute.craigslist.org  19995  2016.0   
423850        terre haute  https://terrehaute.craigslist.org   1400  2002.0   
423851        terre haute  https://terrehaute.craigslist.org   4500  2006.0   
423852        terre haute  https://terrehaute.craigslist.org   4400  1996.0   

       manufacturer                  model  condition    cylinders  ... drive  \
0               gmc        new sierra 1500  excellent  8 cylinders  ...   4wd   
2               gmc  sierra classic 2500hd       good  8 cylinders  ...   4wd   
3              ford                  f-150       good  6 cylinders  ...   4wd   
4         chevrolet                equinox       good  4 cylinders  ...   4wd   
5         chevrolet               traverse       good  6 cylinders  ...   4wd   
...             ...                    ...        ...          ...  ...   ...   
423848         ford                f150 xl       good  6 cylinders  ...   rwd   
423849         mini                 cooper        NaN          NaN  ...   4wd   
423850      mercury       sable ls premium       fair  6 cylinders  ...   fwd   
423851       toyota              sienna le  excellent  6 cylinders  ...   fwd   
423852         ford                f250 xl       fair  8 cylinders  ...   4wd   

             size      type paint_color  \
0             NaN       NaN         red   
2       full-size     truck       white   
3       full-size     truck        grey   
4       full-size       SUV        grey   
5       full-size       SUV       white   
...           ...       ...         ...   
423848  full-size    pickup       brown   
423849        NaN       NaN         NaN   
423850        NaN     sedan         NaN   
423851  full-size  mini-van       brown   
423852  full-size    pickup       white   

                                                image_url  \
0       https://images.craigslist.org/00n0n_f06ykBMcdh...   
2       https://images.craigslist.org/01414_g093aPtSMW...   
3       https://images.craigslist.org/00W0W_io2cwEVtyh...   
4       https://images.craigslist.org/00404_j5GItixnjX...   
5       https://images.craigslist.org/00T0T_bwYCuKicjX...   
...                                                   ...   
423848  https://images.craigslist.org/00J0J_cDy3JSNnQh...   
423849  https://images.craigslist.org/00c0c_5US6NwjpNx...   
423850  https://images.craigslist.org/00707_1IB2ZGpBWf...   
423851  https://images.craigslist.org/00606_isQJg8sRsW...   
423852  https://images.craigs

Now, vehicles with duplicated entries are dropped

In [14]:
vehicles_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 343025 entries, 0 to 423852
Data columns (total 25 columns):
 #   Column        Non-Null Count   Dtype  
---  ------        --------------   -----  
 0   id            343025 non-null  int64  
 1   url           343025 non-null  object 
 2   region        343025 non-null  object 
 3   region_url    343025 non-null  object 
 4   price         343025 non-null  int64  
 5   year          248348 non-null  float64
 6   manufacturer  235869 non-null  object 
 7   model         244910 non-null  object 
 8   condition     144322 non-null  object 
 9   cylinders     152767 non-null  object 
 10  fuel          247262 non-null  object 
 11  odometer      192072 non-null  float64
 12  title_status  247605 non-null  object 
 13  transmission  247898 non-null  object 
 14  vin           103588 non-null  object 
 15  drive         170009 non-null  object 
 16  size          85816 non-null   object 
 17  type          173348 non-null  object 
 18  pain

In [15]:
#Save the new datafame as csv
vehicles_df.to_csv('UsedVehiclesWithNoDuplicatevin.csv', index=False)